---
title: "Analysis of RNA-seq data (tutorial)"
author: "Steve Lianoglou and Thomas Sandmann"
date: "MBL Neurobiology, Summer 2018"
output: 
  rmarkdown::html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    code_folding: hide
vignette: >
  %\VignetteIndexEntry{Analysis of RNA-seq data (tutorial)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
library(dplyr)
library(ggplot2)
theme_set(theme_bw())
source("tutorial-helpers.R")
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Overview

This document is meant to be worked on together in the classroom. We outline
some of the basic steps of a complete RNA-seq analysis by performing the
following:

1. Assembling the raw we have mapped with Kallisto into an analyzable form
2. Performing an initial QC of the dataset
3. Performing your first differential expression analysis
4. Interpreting the result from (3)
5. Using gene sets and gene set enrichment analysis (GSEA) in addition to
   "vanilla" differential expression to assist with the interpretation of
   your experiment.

# Caveats

## There is no joy without suffering

There are sections of this analysis document that are not filled in. This is
intentional, because we will work on these parts together in the class. 

You are meant to struggle through these things so we can:

1. *discover* the answer (no spoon feeding); and
2. perhaps learn more elegant ways to get to it

We understand that this will be very frustrating at times, but trust that:

1. we will get through this together (sleep be damned); and
2. deep down inside, we will be better people for it

## A little joy never hurt nobody

1. For your **future reference**, we have included a sister tutorial in this
   package ("Analysis of RNA-seq data (complete)") which "fills in the holes"
   of this one This can act as a document you can use as reference in the future
   to remind yourself how we performed a "soup to nuts" analysis of this data.
   
   There are other references online that provide workflows for RNA-seq analysis
   that you should also refer to in the future. You will find many examples and
   points of discussion here are liberraly borrow from these sources:
   
   i. [From reads to genes to pathways: differential expression analysis of
       RNA-Seq experiments using Rsubread and the edgeR quasi-likelihood
       pipeline][qlftut]
   ii. [RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR][glimmatut]
   iii. [Analyzing RNA-seq data with DESeq2][deseq2tut]

2. We have provided helper functions in this package to assist with different
   parts of the analysis. These functions start with the `mbl_*` prefix and are
   provided as a crutch so that we can quickly draw certain types of plots, or
   munge certain types of data together.
   
   i.  **We aim to teach you how to make these plots or munge these data 
       without the use of these helper functions by the end of this course**
   ii. These `mbl_*` functions won't disappear, and you can refer (or use)
       them for your future analyses.

3. **One more thing about plotting.** In a later class, we will introduce a
   powerful and (eventually) intuitive system of visualizing data in R using
   the [ggplot2 pakage](http://ggplot2.tidyverse.org). This package provides a
   composable grammar of data visualization that you can use to quickly build
   up arbitrarily detailed plots.
   
   i. There will be code in this tutorial that uses `ggplot` to plot graphics.
      We will explain a bit about this "live," but you will learn more about it
      in a class coming soon.
   ii. the `mbl_*` plotting functions in this package use ggplot2, and you can
      look at that source code as a reference.

# Goal of Differential Expression Analysis

We have performed manipulations to groups of samples (replicates) and want to
understand how these manipulations effect our samples. To measure this effect,
we have chosen to use a "ruler" that measures gene expression estimates of all
of the genes per sample (RNA-seq).

The goal of a differential expression analysis is to identify the **individual**
genes that exhibit **statistically significant** differences in their
**mean expression estimates**.

**Note that** *statistical significance* **does not equal biological
significance**.

## Simple Example

Suppose we want to understand the effect of knocking out an ion channel on the
gene expression profiles taken from Merkel cells. To explore this effect, you
isolate Merkel cells from three WT mice, and three KO mice and perform RNA-seq
on these samples.

We may identify results that land into three broad categories, as outlined
below:

```{r, fig.height = 3, fig.width = 7.6}
exdat <- example_dge_data()
ggplot(exdat, aes(x = condition, y = expression)) +
   geom_boxplot() +
   geom_point() +
   facet_wrap(~ gene)
```

* **Gene A** shows clear differential expression. The mean expression of this
  gene in WT mice is much higher than in KO mice.
* **Gene B** shows questionable differential expression. Although the mean
  gene expression between the two groups is clearly different, there is a
  higher **variance** in these expression estimates, so it's not clear that
  future replicate measurements will show these to be different, or similar.
* **Gene C** the mean expression of these genes between conditions doesn't
  look to be different.

## Formal Assessment of Differential Gene Expression

To address the issues diagrammed above, we will apply statistical tools
developed specifically for RNA-seq data to:

1. estimate the mean differenes in expression between groups of replicate
   samples
2. incorporate the variance in the expression estimates to provide p-values
   for the statistical significance of this mean difference given the this
   variance.

The statistical framework we will use for these analyses is based on
[linear models](https://en.wikipedia.org/wiki/Linear_model).

In your prior encounters with statistics, you may have head of (or used) things
like [t-tests][ttest] or [ANOVA][anova]. Linear models are simply super-charged
versions of these.

# Data Setup

To perform an analysis of gene expression data, we have **three** different
sources of data that we want to merge together.

![Expression Data Schematic](./images/data_schematic.png)

1. The processed RNA-seq data: aka our **assay data**. This consists of a matrix
   of gene-level counts for each sample across our entire experiment. The rows
   of this matrix refer to the individual genes, and the columns refer to the
   individual samples samples. A the value in a given `(x,y)` entry of this
   matrix tells us the number of reads that align to gene `x` in sample `y`.

2. A `data.frame` (table) of **phenotypic data**. The rows of this table refer
   to the different experiments, so there are **as many rows here as there are
   columns in our assay data**. The columns of this table contain the variety of
   covariates we are tracking for these samples, ie. things like:
   
   i. the `"treatment"` condition of the sample;
   ii. the `"genotype"` of the cell(s) (or animal(s)) used in the sample;
   iii. the `"source/tissue"` the sample was taken from;
   iv. the animial identifier for the mouse that sample was taken from
       (you might extract samples from different organs of an animal);
   iv. the person who prepared the libraries for the sample;
   v. the time of day the sample was run;
   vi. the RIN score for the sample;
   vii. etc.

3. A `data.frame` (table) of **gene-level data**. The rows of this table refer
   to the genes quantitated in our experiment, so there are **as many rows here
   as there are rows in our assay data**. The columns provide the information
   we know about these genes, such as:
   
   i. their official [ensembl identifier][ensid]
   ii. official gene name/symbol
   iii. "biotype" (protein coding, lincRNA, etc.)

## Why?

1. Data is easily manipulatable:

   i. indexable/subsettable like a 2d object, you can extract rows and columns
      and the right assay data, phneotypic data, and gene data come along
      for the ride
      
   ii. Different DGE analysis tools use data in this format to find DGE.

2. How does this help do the DGE we outlined above?

## Do it

1. Turn mouse Kallisto quant results into a matrix of gene-level data using
   [tximport][tximport]. Need a vector that points to the `abundance.h5` files
   for each sample.

```{r, eval = FALSE}
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene,
                countsFromAbundance = "lengthScaledTPM")
```

2. TODO: Create DGEList out of everythiing

3. TODO: Redraw ExpressionSet image with `y$counts`, `y$genes`, `y$samples`
   labels.

# Quality Control

In addition to the read-level quality control you already did with FastQC and
MultiQC, there are a number of ways to explore the quality of our data now that
it has been summarized to gene-level expression estimates per sample.

## Effective Sequencing Depth

How many reads aligned to the features (genes) we are analyzing here?

1. TODO: barplot of `lib.size`
2. TODO: barplot of `lib.size` per group

## Principal Component Analysis

TODO: Show simple example of 2d correlated measurements, rotate and project to
1D.

## Heatmaps

TODO: Heatmaps

## QC Conclusion

3. TODO: identify crazy outliers and remove those samples

# Differenial Gene Expression

1. Introduce design matrix
   i. explains experiment (`~ 0 + group`)
   ii. call `lmFit` and show `fit$coefficients`
   iii. `mbl_plot_expression` of genes and show 1:1 match to (ii)
   
2. Introduce contrast as doing arithmetic over columns of the design matrix
3. Run DGE!
4. Introduce alternate paramterization with intercept
   i. show `fit$coefficients`, use `mbl_plot_expression` to eyeball logFC
      calc off of columns vs intercept

5. Use (4) paramaterization to find tissue specific genes
6. Draw heatmap of (5)
   i. use `mbl_heatmap` but explain further customization with `ComplexHeatmap`

# GSEA

multiGSEA reuses what we learned above but gives you much bang for your back

[//]: # References =============================================================

[anova]: https://en.wikipedia.org/wiki/Analysis_of_variance
[deseq2tut]: http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
[ensid]: https://uswest.ensembl.org/Help/Faq?id=488
[glimmatut]: https://f1000research.com/articles/5-1408/v2
[qlftut]: https://f1000research.com/articles/5-1438/v2
[ttest]: https://en.wikipedia.org/wiki/Student%27s_t-test
[tximport]: http://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html

